### PR TITLE
Update Window Locators

### PR DESCRIPTION
**Pull request by titusfortner**
The deprecations are based on the conversation from #849 

The reason for `WindowCollection` is to be able to enforce the deprecation of indexing.

But... Once we have it, we can do some powerful things with it, starting with including `Waitable` similar to how we just did in #857 
Implementing Waitable meant making it relocatable, which is a much better implementation than a simple store of windows.

I have some more ideas for Windows, but this is the next step to make sure on.

### CODE REVIEW

**Comment by titusfortner** [comment]
I just added an element locator to Window because it was so much easier than I thought it would be, and it's a nice replacement if we're taking away index.

File: lib/watir/window_collection.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    end
+
+    alias length count
+    alias size count
+    alias empty? none?
+
+    #
+    # First window of the collection
+    #
+    # @return [Watir::Window] Returns an instance of a Watir::Element subclass
```
Comment (jkotests) [inline lib/watir/window_collection.rb]: Should be `Watir::Window` instead of `Watir::Element`.

File: lib/watir/window_collection.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    #
+
+    def first
+      self[0]
+    end
+
+    #
+    # Last element of the collection
+    #
+    # @return [Watir::Element] Returns an instance of a Watir::Element subclass
```
Comment (jkotests) [inline lib/watir/window_collection.rb]: Both lines should be referring to windows instead of element.

File: lib/watir/window_collection.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    #
+    # Get the element at the given index or range.
+    #
+    # Any call to an ElementCollection that includes an adjacent selector
+    # can not be lazy loaded because it must store the correct type
+    #
+    # Ranges can not be lazy loaded
+    #
+    # @param [Integer, Range] value Index (0-based) or Range of desired element(s)
+    # @return [Watir::Element, Watir::ElementCollection] Returns an instance of a Watir::Element subclass
```
Comment (jkotests) [inline lib/watir/window_collection.rb]: Needs to be updated for windows.

File: lib/watir/window_collection.rb [hunk 4]
```diff
@@ -0,0 +1,98 @@
+module Watir
+  class WindowCollection
+    include Enumerable
+    include Waitable
+
+    def initialize(browser, selector = {})
+      unless selector.keys.all? { |k| %i[title url].include? k }
```
Comment (jkotests) [inline lib/watir/window_collection.rb]: Would it be hard to support `:element` in WindowCollection? It seems useful for Window, so would be nice here too.

**Review by jkotests — CHANGES_REQUESTED** [review]

File: lib/watir/window_collection.rb (hunk 4 above)
Comment (titusfortner) [reply lib/watir/window_collection.rb]: Oh not at all, this was just an oversight; and an indication that I don't have the right tests, either. :/
I squashed in some tests for `ElementCollection#new` and added element to this list.

**Comment by titusfortner** [comment]
It's like I just copy/pasted that other class over and didn't even look at the documentation commends. :)
Should all be fixed now, thanks.

File: lib/watir/window_collection.rb [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      self[-1]
+    end
+
+    #
+    # Get the element at the given index or range.
+    #
+    # @note windows in a collection are not ordered so this is not reliably
+    # @deprecated use Browser#switch_window or a better Window locator
+    # @param [Integer, Range] value Index (0-based) or Range of desired element(s)
+    # @return [Watir::Element, Watir::ElementCollection] Returns an instance of a Watir::Element subclass
```
Comment (jkotests) [inline lib/watir/window_collection.rb]: This documentation still needs an update.

File: lib/watir/window_collection.rb (hunk 4 above)
Comment (titusfortner) [reply lib/watir/window_collection.rb]: and... adding that test showed that my implementation was broken, so... yay tests.

File: lib/watir/window_collection.rb [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    alias eql? ==
+
+    def to_a
+      @to_a ||= begin
+        all = @browser.driver.window_handles.map { |wh| Window.new(@browser, handle: wh) }
+        if @selector.empty?
+          all
+        else
+          all.select do |win|
+            @selector.all? { |key, value| win.send(key) =~ /#{value}/ }
```
Comment (jkotests) [inline lib/watir/window_collection.rb]: As per the spec failure, this doesn't know how to handle the `:element` selector.

**Review by jkotests — CHANGES_REQUESTED** [review]

File: lib/watir/window_collection.rb (hunk 6 above)
Comment (titusfortner) [reply lib/watir/window_collection.rb]: Yeah, took me a couple tries and several more specs to make sure I had this right. :)

File: lib/watir/window_collection.rb [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    # @deprecated use Browser#switch_window or a better Window locator
+    # @return [Watir::Window] Returns an instance of a Watir::Window
+    #
+
+    def last
+      self[-1]
+    end
+
+    #
+    # Get the element at the given index or range.
```
Comment (jkotests) [inline lib/watir/window_collection.rb]: Not to be picky... but still referencing "element" here.

**Review by jkotests — APPROVED** [review]

File: lib/watir/window_collection.rb (hunk 7 above)
Comment (titusfortner) [reply lib/watir/window_collection.rb]: gah. thanks. again.

**Pull request merged**